File: go/tests/wearable_biosignals_test.go
// wearable_biosignals_test.go
//
// Verifies the CircleAI.Wearable.Biosignals port (wearable_biosignals.go):
// sample factory (confidence clamp), null + recorded sources, sliding-window
// aggregation stats, and the deterministic affect-mapper rule sheet.

package circleai_test

import (
	"context"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestBiosignals_SampleFactoryAndSources(t *testing.T) {
	s := circleai.NewBiosignalSample(circleai.BiosignalKindHeartRate, 72, "bpm", 1.5, false)
	if s.Confidence != 1.0 {
		t.Fatalf("confidence must clamp to 1.0, got %v", s.Confidence)
	}
	if s.Kind != circleai.BiosignalKindHeartRate || s.Value != 72 || s.Unit != "bpm" {
		t.Fatalf("sample fields wrong: %+v", s)
	}

	// Null source: supports nothing, streams nothing.
	var null circleai.BiosignalSource = circleai.NullBiosignalSource{}
	if len(null.SupportedKinds()) != 0 || null.IsSupported(circleai.BiosignalKindHeartRate) {
		t.Fatalf("null source must support nothing")
	}
	count := 0
	for range null.Stream(context.Background()) {
		count++
	}
	if count != 0 {
		t.Fatalf("null source must stream nothing, got %d", count)
	}

	// Recorded source replays its samples and reports its kinds.
	samples := []circleai.BiosignalSample{
		circleai.NewBiosignalSample(circleai.BiosignalKindHeartRate, 60, "bpm", 1, false),
		circleai.NewBiosignalSample(circleai.BiosignalKindSteps, 100, "count", 1, true),
	}
	rec, err := circleai.NewRecordedBiosignalSource(samples, 0)
	if err != nil {
		t.Fatalf("new recorded: %v", err)
	}
	if !rec.IsSupported(circleai.BiosignalKindHeartRate) || rec.IsSupported(circleai.BiosignalKindOxygenSaturation) {
		t.Fatalf("recorded IsSupported wrong")
	}
	got := 0
	for range rec.Stream(context.Background()) {
		got++
	}
	if got != 2 {
		t.Fatalf("recorded stream count = %d, want 2", got)
	}
	if _, err := circleai.NewRecordedBiosignalSource(nil, 0); err == nil {
		t.Fatalf("nil samples must error")
	}
}

func TestBiosignals_AggregatorSnapshot(t *testing.T) {
	now := time.Now().UTC()
	mk := func(k circleai.BiosignalKind, v float32) circleai.BiosignalSample {
		s := circleai.NewBiosignalSample(k, v, "u", 1, false)
		s.MeasuredAt = now
		return s
	}
	rec, _ := circleai.NewRecordedBiosignalSource([]circleai.BiosignalSample{
		mk(circleai.BiosignalKindHeartRate, 60),
		mk(circleai.BiosignalKindHeartRate, 80),
		mk(circleai.BiosignalKindHeartRate, 100),
	}, 0)
	agg, err := circleai.NewBiosignalAggregator(rec)
	if err != nil {
		t.Fatalf("new aggregator: %v", err)
	}
	snap, err := agg.Snapshot(context.Background(), time.Minute)
	if err != nil {
		t.Fatalf("snapshot: %v", err)
	}
	hr, ok := snap.Stats[circleai.BiosignalKindHeartRate]
	if !ok {
		t.Fatalf("no HeartRate stats in snapshot: %+v", snap.Stats)
	}
	if hr.SampleCount != 3 || hr.Min != 60 || hr.Max != 100 || hr.Mean != 80 {
		t.Fatalf("HR stats wrong: %+v", hr)
	}
	if _, err := agg.Snapshot(context.Background(), 0); err == nil {
		t.Fatalf("non-positive window must error")
	}
}

func TestBiosignals_AffectMapper(t *testing.T) {
	// High heart rate raises Energy and Uncertainty.
	a := circleai.NewAffectState("u1")
	e0, u0 := a.Energy, a.Uncertainty
	hr := circleai.NewBiosignalSample(circleai.BiosignalKindHeartRate, 140, "bpm", 1, false)
	circleai.ApplyBiosignalToAffect(hr, &a)
	if a.Energy <= e0 || a.Uncertainty <= u0 {
		t.Fatalf("high HR should raise Energy and Uncertainty: e %v->%v u %v->%v", e0, a.Energy, u0, a.Uncertainty)
	}

	// Low confidence is ignored.
	b := circleai.NewAffectState("u2")
	eb := b.Energy
	lowConf := circleai.NewBiosignalSample(circleai.BiosignalKindHeartRate, 140, "bpm", 0.2, false)
	circleai.ApplyBiosignalToAffect(lowConf, &b)
	if b.Energy != eb {
		t.Fatalf("low-confidence sample must not mutate affect: %v -> %v", eb, b.Energy)
	}

	// Low SpO2 raises Uncertainty.
	c := circleai.NewAffectState("u3")
	uc := c.Uncertainty
	spo2 := circleai.NewBiosignalSample(circleai.BiosignalKindOxygenSaturation, 85, "%", 1, false)
	circleai.ApplyBiosignalToAffect(spo2, &c)
	if c.Uncertainty <= uc {
		t.Fatalf("low SpO2 should raise Uncertainty: %v -> %v", uc, c.Uncertainty)
	}

	// SleepStage does not mutate the affect dimensions.
	d := circleai.NewAffectState("u4")
	ed, ud := d.Energy, d.Uncertainty
	sleep := circleai.NewBiosignalSample(circleai.BiosignalKindSleepStage, 2, "stage", 1, false)
	circleai.ApplyBiosignalToAffect(sleep, &d)
	if d.Energy != ed || d.Uncertainty != ud {
		t.Fatalf("sleep stage must not change affect dimensions")
	}
}

File: rust/tests/wearable_biosignals_test.rs
//! wearable_biosignals_test.rs
//!
//! Ports the behaviour of `CircleAI.Wearable.Biosignals`: sample factory
//! (confidence clamp + fresh id), recorded/null sources, sliding-window
//! aggregator stats, and the deterministic affect mapper rule sheet.

use chrono::{Duration, Utc};
use circle_ai::memory::AffectState;
use circle_ai::wearable_biosignals::{
    BiosignalAffectMapper, BiosignalAggregator, BiosignalKind, BiosignalSample, IBiosignalSource,
    NullBiosignalSource, RecordedBiosignalSource,
};

#[test]
fn kind_integer_values_are_stable() {
    assert_eq!(BiosignalKind::HeartRate as i32, 0);
    assert_eq!(BiosignalKind::HeartRateVariability as i32, 1);
    assert_eq!(BiosignalKind::OxygenSaturation as i32, 2);
    assert_eq!(BiosignalKind::Accelerometer as i32, 3);
    assert_eq!(BiosignalKind::BodyTemperature as i32, 4);
    assert_eq!(BiosignalKind::SleepStage as i32, 5);
    assert_eq!(BiosignalKind::Steps as i32, 6);
    assert_eq!(BiosignalKind::GalvanicSkinResponse as i32, 7);
    assert_eq!(BiosignalKind::Unknown as i32, 8);
}

#[test]
fn create_clamps_confidence_and_assigns_id() {
    let s = BiosignalSample::create(BiosignalKind::HeartRate, 72.0, "bpm", 2.5, false);
    assert_eq!(s.confidence, 1.0); // clamped to [0,1]
    assert!(!s.id.is_nil());
    let s2 = BiosignalSample::create(BiosignalKind::HeartRate, 72.0, "bpm", -1.0, false);
    assert_eq!(s2.confidence, 0.0);
    assert_ne!(s.id, s2.id); // fresh id each time
}

#[test]
fn null_source_emits_nothing() {
    let src = NullBiosignalSource::new();
    assert!(src.supported_kinds().is_empty());
    assert!(src.stream().is_empty());
    assert!(!src.is_supported(BiosignalKind::HeartRate));
}

#[test]
fn recorded_source_reports_distinct_kinds_and_replays() {
    let now = Utc::now();
    let samples = vec![
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 70.0, "bpm", 1.0, false, now),
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 72.0, "bpm", 1.0, false, now),
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::Steps, 100.0, "count", 1.0, true, now),
    ];
    let src = RecordedBiosignalSource::new(samples, None);
    assert_eq!(src.stream().len(), 3);
    assert_eq!(src.supported_kinds().len(), 2); // HeartRate + Steps
    assert!(src.is_supported(BiosignalKind::Steps));
    assert!(!src.is_supported(BiosignalKind::OxygenSaturation));
    assert_eq!(src.replay_delay(), Duration::zero());
}

#[test]
fn aggregator_computes_min_max_mean_within_window() {
    let now = Utc::now();
    let samples = vec![
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 60.0, "bpm", 1.0, false, now),
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 80.0, "bpm", 1.0, false, now),
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 100.0, "bpm", 1.0, false, now),
        // Outside the window (measured well before cutoff) — excluded.
        BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 999.0, "bpm", 1.0, false, now - Duration::hours(2)),
    ];
    let src = RecordedBiosignalSource::new(samples, None);
    let agg = BiosignalAggregator::new(&src);
    let snap = agg.snapshot(Duration::minutes(5));
    let stats = snap.stats.get(&BiosignalKind::HeartRate).unwrap();
    assert_eq!(stats.sample_count, 3);
    assert_eq!(stats.min, 60.0);
    assert_eq!(stats.max, 100.0);
    assert!((stats.mean - 80.0).abs() < 1e-4);
}

#[test]
#[should_panic(expected = "Window must be positive.")]
fn aggregator_zero_window_panics() {
    let src = NullBiosignalSource::new();
    let agg = BiosignalAggregator::new(&src);
    agg.snapshot(Duration::zero());
}

#[test]
fn affect_mapper_high_heart_rate_raises_energy_and_uncertainty() {
    let mut affect = AffectState::new("u");
    affect.energy = 0.5;
    affect.uncertainty = 0.2;
    let sample = BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 140.0, "bpm", 1.0, false, Utc::now());
    BiosignalAffectMapper::apply(&sample, &mut affect);
    assert!((affect.energy - 0.60).abs() < 1e-4);
    assert!((affect.uncertainty - 0.25).abs() < 1e-4);
}

#[test]
fn affect_mapper_low_confidence_is_ignored() {
    let mut affect = AffectState::new("u");
    affect.energy = 0.5;
    let sample = BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::HeartRate, 140.0, "bpm", 0.4, false, Utc::now());
    BiosignalAffectMapper::apply(&sample, &mut affect);
    assert!((affect.energy - 0.5).abs() < 1e-9); // unchanged
}

#[test]
fn affect_mapper_low_spo2_raises_uncertainty() {
    let mut affect = AffectState::new("u");
    affect.uncertainty = 0.2;
    let sample = BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::OxygenSaturation, 88.0, "%", 1.0, false, Utc::now());
    BiosignalAffectMapper::apply(&sample, &mut affect);
    assert!((affect.uncertainty - 0.30).abs() < 1e-4);
}

#[test]
fn affect_mapper_sleep_stage_no_mutation() {
    let mut affect = AffectState::new("u");
    affect.energy = 0.5;
    affect.uncertainty = 0.2;
    let sample = BiosignalSample::new(uuid::Uuid::new_v4(), BiosignalKind::SleepStage, 2.0, "stage", 1.0, false, Utc::now());
    BiosignalAffectMapper::apply(&sample, &mut affect);
    assert!((affect.energy - 0.5).abs() < 1e-9);
    assert!((affect.uncertainty - 0.2).abs() < 1e-9);
}
